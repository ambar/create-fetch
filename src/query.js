import querystring from 'querystring'

/**
 * Stringify query string
 */
const query = () => fetch => (url, {query, ...options} = {}) => {
  if (typeof query === 'object') {
    const params = querystring.stringify(query)
    if (params) {
      const index = url.lastIndexOf('?')
      url += (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
    }
  }

  return fetch(url, options)
}

export default query
