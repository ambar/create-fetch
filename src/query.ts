import getEntries from './utils/getEntries'
import {FetchEnhancer} from './types'

// preserve legacy querystring/qs behavior, set null or undefined to empty
const mapNilToEmpty = (obj: Record<string, any>) =>
  getEntries(obj).map(([k, v]) => [k, (v as string) ?? ''])

type QueryInit = {
  query?: string[][] | Record<string, any> | URLSearchParams
}

/**
 * Stringify query string
 */
const query =
  (): FetchEnhancer<QueryInit> =>
  (fetch) =>
  (url, {query, ...options} = {}) => {
    if (query !== null && typeof query === 'object') {
      const params = (
        query instanceof URLSearchParams
          ? query
          : new URLSearchParams(mapNilToEmpty(query))
      ).toString()
      if (params) {
        url = String(url)
        const index = url.lastIndexOf('?')
        url +=
          (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
      }
    }

    return fetch(url, options)
  }

export default query
