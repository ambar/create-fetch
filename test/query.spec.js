import query from '../src/query'

describe('query', () => {
  const fetch = url => Promise.resolve(url)

  it('should add query string', async () => {
    const myFetch = query()(fetch)
    expect(await myFetch('/api', {query: {foo: 1, bar: 2}})).toBe(
      '/api?foo=1&bar=2'
    )
    expect(
      await myFetch('/api', {query: new URLSearchParams({foo: 'bar'})})
    ).toBe('/api?foo=bar')
  })

  it('should not add query string with empty query', async () => {
    const myFetch = query()(fetch)
    expect(await myFetch('/api', {query: {}})).toBe('/api')
    expect(await myFetch('/api', {query: null})).toBe('/api')
  })

  it('should handle existing query string', async () => {
    const myFetch = query()(fetch)
    expect(await myFetch('/api?ok', {query: {foo: 1}})).toBe('/api?ok&foo=1')
    expect(await myFetch('/api?', {query: {foo: 1}})).toBe('/api?foo=1')
    expect(await myFetch('/api?', {query: {}})).toBe('/api?')
    expect(await myFetch('/api?', {query: false})).toBe('/api?')
  })

  it('should work without options', async () => {
    const myFetch = query()(fetch)
    expect(await myFetch('/api')).toMatchSnapshot()
  })
})
