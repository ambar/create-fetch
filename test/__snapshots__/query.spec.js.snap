// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query should work without options 1`] = `"/api"`;
